File: src/utils/chiyototatsu.cpp
#include <ctime>

#include "../compiled_proto/backup_category.pb.h"
#include "../compiled_proto/tachiyomi.pb.h"

#include "chiyotatsu_util.hpp"
#include "kotatsu_utils.hpp"
#include "tachiyomi_utils.hpp"

using namespace kotatsu;
using namespace std;

namespace chiyotatsu
{
void chiyototatsu(const Backup *tachiyomiBackup, KotatsuBackup *reference)
{
    // (To be copied over if there's a reference file)
    // Index, History, Bookmarks, Sources, Settings
    // Categories (kotatsu) + categories (tachiyomi, prepended with chtsu)
    // Favourites (kotatsu) + MangaBackups (tachiyomi)

    // Append categories
    int64_t origCatCount = reference->categories.size();
    for (auto tCat : tachiyomiBackup->categories()) {
        Categories kCat;
        kCat.created_at  = time(NULL);
        kCat.category_id = origCatCount + tCat.order();
        kCat.title       = "CHTSU - " + tCat.name();

        // Rando default values
        kCat.show_in_lib = true;
        kCat.track       = false;
        kCat.order       = "NEWEST";
        kCat.sort_key    = reference->categories.size();

        reference->categories.push_back(kCat);
    }

    for (auto tManga : tachiyomiBackup->mangabackup()) {
        Favourites kManga;

        kManga.manga.title  = tManga.title();
        kManga.manga.author = tManga.author();

        tachiyomi::TachiyomiSource source =
            tachiyomi::PBoneSourceIDs.at(tManga.source());
        transform(
            source.name.begin(), source.name.end(), source.name.begin(),
            ::toupper
        );
        kManga.manga.source    = source.name;
        kManga.manga.url       = source.baseUrl + tManga.url();
        kManga.manga.cover_url = source.baseUrl + tManga.thumbnailurl();

        // Rando values again
        kManga.manga.nsfw      = false;
        kManga.manga.rating    = 4.20F;
        kManga.manga.tags      = {};
        kManga.manga.id        = 0;
        kManga.manga.alt_title = "";

        // Rando values again
        kManga.category_id =
            reference->categories.size() - 1; // Last one for now
        kManga.created_at  = time(NULL);
        kManga.manga_id    = 0;
        kManga.sort_key    = 0;

        // tManga.status(); I dunno what this enum turns out to be
        kManga.manga.state = "";

        reference->favourites.push_back(kManga);
    }
}
} // namespace chiyotatsu

File: utils/chiyototatsu/chiyototatsu.go
package chiyototatsu

import (
	"strings"
	"time"

	"github.com/croisen/chiyo-tatsu/utils/kotatsu_backups"
	"github.com/croisen/chiyo-tatsu/utils/tachiyomi_backups"
)

func Chiyototatsu(tBackup *tachiyomi_backups.Backup, kBackup *kotatsu_backups.Backup) *kotatsu_backups.Backup {
	kCats := kBackup.Categories
	kCatDef := kotatsu_backups.Category{
		Category_id: uint64(len(kCats)) + 1,
		Created_at:  uint64(time.Now().Unix()),
		Sort_key:    uint64(len(kCats)) + 1,
		Title:       "CH - Default",
		Order:       "NEWEST", // Rando up to the end
		Track:       false,
		Show_in_lib: false,
	}

	kCats = append(kCats, kCatDef)
	kCatSize := uint64(len(kCats))
	for _, tCat := range tBackup.Categories {
		kCat := kotatsu_backups.Category{
			Category_id: kCatSize + uint64(tCat.Order),
			Created_at:  uint64(time.Now().Unix()),
			Sort_key:    kCatSize + uint64(tCat.Order),
			Title:       "CH - " + tCat.Name,

			Order:       "NEWEST", // Rando up to the end
			Track:       false,
			Show_in_lib: false,
		}
		kCats = append(kCats, kCat)
	}

	kBackup.Categories = kCats
	for _, tManga := range tBackup.MangaBackup {
		catID := kCatSize
		if len(tManga.GetCategories()) != 0 {
			catID = kCatSize + uint64(tManga.Categories[0])
		}

		kFav := kotatsu_backups.Favourite{
			Category_id: catID,
			Created_at:  uint64(time.Now().Unix()),

			Manga_id: 0, // Rando again
			Sort_key: 0,
		}

		kManga := kotatsu_backups.Manga{
			Title:     tManga.GetTitle(),
			Alt_title: tManga.GetTitle(),

			Cover_url:  tManga.GetThumbnailUrl(),
			Url:        "", // tManga.GetUrl(), TODO yet
			Public_url: "", // tManga.GetUrl(), TODO yet

			Source: "", // TODO

			Id:     0, // Rando up to the end for now
			Nsfw:   false,
			Rating: 4.2,
			State:  "",
		}

		for _, tGenre := range tManga.GetGenre() {
			kTag := kotatsu_backups.Tags{
				Title:  tGenre,
				Key:    strings.ToLower(tGenre),
				Id:     0,
				Source: "", // TODO
			}

			kManga.Tags = append(kManga.Tags, kTag)
		}

		kFav.Manga = kManga
		kBackup.Favourites = append(kBackup.Favourites, kFav)
	}

	return kBackup
}
